integration-tests/src/test/java/com/datastax/oss/driver/api/core/cql/PerRequestKeyspaceIT.java

```diff
@@ -203,4 +204,33 @@ public void should_prepare_statement_with_keyspace() {
             .one();
     assertThat(row.getInt(0)).isEqualTo(1);
   }
+
+  @Test
+  @CassandraRequirement(min = "4.0")
+  public void should_reprepare_statement_with_keyspace_on_the_fly() {
+    // Create a separate session because we don't want it to have a default keyspace
+    try (CqlSession session = SessionUtils.newSession(ccmRule)) {
+      executeDdl(
+          session,
+          String.format(
+              "CREATE TABLE IF NOT EXISTS %s.bar (k int primary key)", sessionRule.keyspace()));
+      PreparedStatement pst =
+          session.prepare(
+              SimpleStatement.newInstance("SELECT * FROM bar WHERE k=?")
+                  .setKeyspace(sessionRule.keyspace()));
+
+      // Drop and re-create the table to invalidate the prepared statement server side
+      executeDdl(session, String.format("DROP TABLE %s.bar", sessionRule.keyspace()));
+      executeDdl(
+          session,
+          String.format("CREATE TABLE %s.bar (k int primary key)", sessionRule.keyspace()));
+
+      // This will re-prepare on the fly
+      session.execute(pst.bind(0));
```

Comment:
maybe we can check the cache to see that it was re-prepared and populated?

Reply:
I assume you mean the server-side cache. I can check `system.prepared_statements`, it will always be present since the test target C* 4+.